openedx/openedx-platform
# Soft delete collections

## Description

Adds support for soft-deleting and restoring Collections.

As part of this change, we're refactoring the way the Collections-related events are triggered to use Django model signals, so that changes made in Django Admin will be reflected in the Studio search index.

- Which edX user roles will this change impact? Library Authors; not enabled in production by default.

## Supporting information

Related to: https://github.com/openedx/modular-learning/issues/231
Depends on: https://github.com/openedx/openedx-learning/pull/229
Private-ref: [FAL-3809](https://tasks.opencraft.com/browse/FAL-3809)

## Testing instructions

Recommend reviewing the code commit by commit.

1. Reindex your studio content to update any existing Collection documents: `tutor dev run cms ./manage.py cms reindex_studio --experimental`
2. In the Authoring frontend, create a Collection and add some components to it.
3. Open http://meilisearch.local.edly.io:7700/
    Your api key can be found with tutor config printvalue MEILISEARCH_API_KEY
    Search for your collection key and verify that the collection document and its components are returned.
4. Hit the "soft delete" REST API: DELETE http://studio.local.edly.io:8001/api/libraries/v2/<library_key>/collections/<collection_key>/
    Note that the collection no longer appears in the Authoring MFE list of collections.
    Search for your collection key in Meilisearch and verify that no results are found.
5. Hit the "restore" REST API: POST http://studio.local.edly.io:8001/api/libraries/v2/<library_key>/collections/<collection_key>/restore/
   Verify that the collection again appears in the Authoring MFE list of collections.
    Search for your collection key in Meilisearch and verify that the collection document and its components are returned again.
6. Open *Studio* Django Admin http://studio.local.edly.io:8001/admin/oel_collections/collection/
     Disabling a collection should have the same effect as the "soft delete" REST API (step 4).
     Re-enabling a collection should have the same effect as the "restore" REST API (step 5)
7. Delete a Collection via Django Admin.
     This will permanently delete the Collection.
8. Try to hit the "restore" REST API with the deleted collection key: POST http://studio.local.edly.io:8001/api/libraries/v2/<library_key>/collections/<collection_key>/restore/
   This should return a 404.

## Deadline

ASAP

## Author Notes & Concerns

* f760c2b6d561093e163a2d92794578d30b9d8726 changes the Meilisearch ID used for Collections from the int `Collection.pk` to a string generated from the Collection's usage key.
  We needed to do this so that the Collection document could be identified by the information in the `CONTENT_LIBRARY_DELETED` event (`library_key` + `collection_key`), so this document can be removed after the Collection itself is removed.
  Users will need to reindex their studio search content.
